jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java

```diff
@@ -596,7 +598,40 @@ default Enumeration<String> getFieldNames()
      */
     default Set<String> getFieldNamesCollection()
     {
-        return stream().map(HttpField::getName).collect(Collectors.toSet());
+        Set<HttpHeader> seenByHeader = EnumSet.noneOf(HttpHeader.class);
+        Set<String> seenByName = null;
+        List<String> list = new ArrayList<>(size());
```

Comment:
Can't we just allocate a `LinkedHashSet` and return that directly instead of wrapping with an `AbstractSet`, as the `LinkedHashSet` will retain the insertion order.

Reply:
Two problems with that approach: 
 1. LinkedHashSet doesn't do a case insensitive key
 2. The intention is that most/all headers will be of a known type and never be put into the set and only flagged in the more efficient EnumSet.

Possibly we could create a TreeSet comparator that used the existing order... but that would be more complex.

Reply:
I meant just replacing the `ArrayList` with `LinkedHashSet`, instead of having `AbstractSet` backed by an `ArrayList`. You would still use the `TreeSet` for `seenByName`.

Reply:
Ah OK.  However I think the AbstractSet back by a list is still better, as this set is primarily used for iteration rather than queries. So there is no need to build the hash lookup mechanism.